File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/util/RequestModelUtilsTest.java
package com.emarsys.mobileengage.util;

import com.emarsys.core.device.DeviceInfo;
import com.emarsys.core.device.LanguageProvider;
import com.emarsys.core.provider.hardwareid.HardwareIdProvider;
import com.emarsys.core.provider.timestamp.TimestampProvider;
import com.emarsys.core.provider.uuid.UUIDProvider;
import com.emarsys.core.request.model.RequestMethod;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.core.util.TimestampUtils;
import com.emarsys.mobileengage.RequestContext;
import com.emarsys.mobileengage.event.applogin.AppLoginParameters;
import com.emarsys.mobileengage.storage.AppLoginStorage;
import com.emarsys.mobileengage.storage.MeIdSignatureStorage;
import com.emarsys.mobileengage.storage.MeIdStorage;
import com.emarsys.mobileengage.testUtil.RequestModelTestUtils;
import com.emarsys.testUtil.InstrumentationRegistry;
import com.emarsys.testUtil.SharedPrefsUtils;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static junit.framework.Assert.assertFalse;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class RequestModelUtilsTest {
    private static final String APPLICATION_CODE = "applicationCode";
    private static final String APPLICATION_PASSWORD = "applicationPassword";
    public static final String VALID_CUSTOM_EVENT_V3 = "https://mobile-events.eservice.emarsys.net/v3/devices/12345/events";
    public static final String REQUEST_ID = "REQUEST_ID";
    public static final String SDK_VERSION = "sdkVersion";
    public static final long TIMESTAMP = 100_000;

    private RequestContext requestContext;
    private MeIdStorage meIdStorage;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();
    private UUIDProvider uuidProvider;
    private TimestampProvider timestampProvider;

    @Before
    public void setup() {
        SharedPrefsUtils.clearSharedPrefs("emarsys_shared_preferences");

        meIdStorage = mock(MeIdStorage.class);
        uuidProvider = mock(UUIDProvider.class);
        when(uuidProvider.provideId()).thenReturn(REQUEST_ID);

        timestampProvider = mock(TimestampProvider.class);
        when(timestampProvider.provideTimestamp()).thenReturn(TIMESTAMP);
        requestContext = new RequestContext(
                APPLICATION_CODE,
                APPLICATION_PASSWORD,
                1,
                new DeviceInfo(InstrumentationRegistry.getTargetContext(),
                        mock(HardwareIdProvider.class),
                        SDK_VERSION,
                        mock(LanguageProvider.class)),
                mock(AppLoginStorage.class),
                meIdStorage,
                mock(MeIdSignatureStorage.class),
                timestampProvider,
                uuidProvider);

        requestContext.setAppLoginParameters(new AppLoginParameters(3, "test@test.com"));
    }

    @After
    public void tearDown() {
        SharedPrefsUtils.clearSharedPrefs("emarsys_shared_preferences");
    }

    @Test(expected = IllegalArgumentException.class)
    public void testIsCustomEvent_V3_mustNotBeNull() {
        RequestModelUtils.isCustomEvent_V3(null);
    }

    @Test
    public void testIsCustomEvent_V3_returnsTrue_ifIndeedV3Event() {
        RequestModel requestModel = new RequestModel.Builder(timestampProvider, uuidProvider)
                .url(VALID_CUSTOM_EVENT_V3)
                .build();

        assertTrue(RequestModelUtils.isCustomEvent_V3(requestModel));
    }

    @Test
    public void testIsCustomEvent_V3_returnsFalse_ifThereIsNoMatch() {
        RequestModel requestModel = new RequestModel.Builder(timestampProvider, uuidProvider)
                .url("https://www.google.com")
                .build();

        assertFalse(RequestModelUtils.isCustomEvent_V3(requestModel));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateAppLogin_V2_requestContext_mustNotBeNull() {
        RequestModelUtils.createAppLogin_V2(null, "pushtoken");
    }

    @Test
    public void testCreateAppLogin_V2() {
        RequestModel expected = new RequestModel.Builder(timestampProvider, uuidProvider)
                .url("https://push.eservice.emarsys.net/api/mobileengage/v2/users/login")
                .payload(RequestPayloadUtils.createAppLoginPayload(requestContext, null))
                .headers(RequestHeaderUtils.createBaseHeaders_V2(requestContext))
                .build();

        RequestModel result = RequestModelUtils.createAppLogin_V2(requestContext, null);

        RequestModelTestUtils.assertEqualsRequestModels(expected, result);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateLastMobileActivity_requestContext_mustNotBeNull() {
        RequestModelUtils.createLastMobileActivity(null);
    }

    @Test
    public void testCreateLastMobileActivity_V3() {
        when(meIdStorage.get()).thenReturn("meId");

        RequestModel expected = new RequestModel.Builder(timestampProvider, uuidProvider)
                .url(RequestUrlUtils.createEventUrl_V3(requestContext.getMeIdStorage().get()))
                .payload(RequestPayloadUtils.createBasePayload(requestContext))
                .headers(RequestHeaderUtils.createBaseHeaders_V3(requestContext))
                .build();

        RequestModel requestModel = RequestModelUtils.createLastMobileActivity(requestContext);

        assertEquals(expected.getUrl(), requestModel.getUrl());
        assertEquals("last_mobile_activity", (((List<Map<String, Object>>) requestModel.getPayload().get("events")).get(0).get("name")));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateInternalCustomEvent_eventNameShouldNotBeNull() {
        RequestModelUtils.createInternalCustomEvent(
                null,
                new HashMap<String, String>(),
                requestContext
        );
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateInternalCustomEvent_requestContextShouldNotBeNull() {
        RequestModelUtils.createInternalCustomEvent(
                "eventname",
                new HashMap<String, String>(),
                null
        );
    }

    @Test
    public void testCreateInternalCustomEvent_withoutAttributes() {
        long timestamp = 90_000;
        when(requestContext.getTimestampProvider().provideTimestamp()).thenReturn(timestamp);
        String eventName = "name";
        String meId = "12345";
        String meIdSignature = "12345";
        when(requestContext.getMeIdStorage().get()).thenReturn(meId);
        when(requestContext.getMeIdSignatureStorage().get()).thenReturn(meIdSignature);

        Map<String, Object> event = new HashMap<>();
        event.put("type", "internal");
        event.put("name", eventName);
        event.put("timestamp", TimestampUtils.formatTimestampWithUTC(timestamp));

        Map<String, Object> payload = new HashMap<>();
        payload.put("clicks", new ArrayList<>());
        payload.put("viewed_messages", new ArrayList<>());
        payload.put("events", Collections.singletonList(event));

        RequestModel actual = RequestModelUtils.createInternalCustomEvent(
                eventName,
                null,
                requestContext);

        RequestModel expected = new RequestModel(
                RequestUrlUtils.createEventUrl_V3(meId),
                RequestMethod.POST,
                payload,
                RequestHeaderUtils.createBaseHeaders_V3(requestContext),
                timestamp,
                Long.MAX_VALUE,
                actual.getId());

        assertEquals(expected, actual);
    }

    @Test
    public void testCreateInternalCustomEvent_withAttributes() {
        long timestamp = 90_000;
        when(requestContext.getTimestampProvider().provideTimestamp()).thenReturn(timestamp);
        String eventName = "name";
        String meId = "12345";
        String meIdSignature = "12345";
        when(requestContext.getMeIdStorage().get()).thenReturn(meId);
        when(requestContext.getMeIdSignatureStorage().get()).thenReturn(meIdSignature);

        Map<String, String> attributes = new HashMap<>();
        attributes.put("key1", "value1");
        attributes.put("key2", "value2");
        attributes.put("key3", "value3");

        Map<String, Object> event = new HashMap<>();
        event.put("type", "internal");
        event.put("name", eventName);
        event.put("timestamp", TimestampUtils.formatTimestampWithUTC(timestamp));
        event.put("attributes", attributes);

        Map<String, Object> payload = new HashMap<>();
        payload.put("clicks", new ArrayList<>());
        payload.put("viewed_messages", new ArrayList<>());
        payload.put("events", Collections.singletonList(event));

        RequestModel actual = RequestModelUtils.createInternalCustomEvent(
                eventName,
                attributes,
                requestContext);

        RequestModel expected = new RequestModel(
                RequestUrlUtils.createEventUrl_V3(meId),
                RequestMethod.POST,
                payload,
                RequestHeaderUtils.createBaseHeaders_V3(requestContext),
                timestamp,
                Long.MAX_VALUE,
                actual.getId());

        assertEquals(expected, actual);
    }
}

File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/util/RequestModelUtilsTest.kt
package com.emarsys.mobileengage.util

import android.os.Handler
import android.os.Looper
import com.emarsys.core.di.DependencyInjection
import com.emarsys.core.di.getDependency
import com.emarsys.core.endpoint.ServiceEndpointProvider
import com.emarsys.core.request.model.RequestModel
import com.emarsys.mobileengage.fake.FakeMobileEngageDependencyContainer
import com.emarsys.mobileengage.util.RequestModelUtils.isCustomEvent
import com.emarsys.mobileengage.util.RequestModelUtils.isMobileEngageRequest
import com.emarsys.mobileengage.util.RequestModelUtils.isRefreshContactTokenRequest
import com.emarsys.mobileengage.util.RequestModelUtils.isRemoteConfigRequest
import com.emarsys.testUtil.TimeoutUtils
import com.emarsys.testUtil.mockito.whenever
import io.kotlintest.shouldBe
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.mockito.Mockito.mock
import java.net.URL

class RequestModelUtilsTest {

    private companion object {
        const val CLIENT_HOST = "https://me-client.eservice.emarsys.net/v3"
        const val CLIENT_BASE = "$CLIENT_HOST/apps/%s/client"
        const val EVENT_HOST = "https://mobile-events.eservice.emarsys.net"
        const val EVENT_BASE = "$EVENT_HOST/v3/apps/%s/client/events"
        const val EVENT_BASE_V4 = "$EVENT_HOST/v4/apps/%s/client/events"
        const val INBOX_HOST = "https://mobile-events.eservice.emarsys.net/v3"
        const val INBOX_BASE = "$INBOX_HOST/apps/%s/inbox"
        const val REMOTE_CONFIG_HOST = "https://mobile-sdk-config.gservice.emarsys.net"
    }

    private lateinit var mockClientServiceProvider: ServiceEndpointProvider
    private lateinit var mockEventServiceProvider: ServiceEndpointProvider
    private lateinit var mockMessageInboxServiceProvider: ServiceEndpointProvider
    private lateinit var mockRequestModel: RequestModel

    @Before
    fun setUp() {
        mockClientServiceProvider = mock(ServiceEndpointProvider::class.java).apply {
            whenever(provideEndpointHost()).thenReturn(CLIENT_HOST)
        }
        mockEventServiceProvider = mock(ServiceEndpointProvider::class.java).apply {
            whenever(provideEndpointHost()).thenReturn(EVENT_HOST)
        }
        mockMessageInboxServiceProvider = mock(ServiceEndpointProvider::class.java).apply {
            whenever(provideEndpointHost()).thenReturn(INBOX_HOST)
        }

        mockRequestModel = mock(RequestModel::class.java)
        DependencyInjection.setup(
                FakeMobileEngageDependencyContainer(
                        clientServiceProvider = mockClientServiceProvider,
                        eventServiceProvider = mockEventServiceProvider,
                        messageInboxServiceProvider = mockMessageInboxServiceProvider
                ))
    }

    @After
    fun tearDown() {
        val handler = getDependency<Handler>("coreSdkHandler")
        val looper: Looper? = handler.looper
        looper?.quit()
        DependencyInjection.tearDown()
    }

    @Rule
    @JvmField
    val timeout: TestRule = TimeoutUtils.timeoutRule

    @Test
    fun testIsMobileEngageRequest_true_whenItIsMobileEngageClient() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL(CLIENT_BASE))
        }
        val result = mockRequestModel.isMobileEngageRequest()

        result shouldBe true
    }

    @Test
    fun testIsMobileEngageRequest_true_whenItIsMobileEngageEvent() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL(EVENT_BASE))
        }
        val result = mockRequestModel.isMobileEngageRequest()

        result shouldBe true
    }

    @Test
    fun testIsMobileEngageRequest_true_whenItIsMobileEngageEventV4() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL(EVENT_BASE_V4))
        }
        val result = mockRequestModel.isMobileEngageRequest()

        result shouldBe true
    }

    @Test
    fun testIsMobileEngageRequest_true_whenItIsMobileEngageMessageInbox() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL(INBOX_BASE))
        }
        val result = mockRequestModel.isMobileEngageRequest()

        result shouldBe true
    }

    @Test
    fun testIsMobileEngageRequest_false_whenItIsNotMobileEngage() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL("https://not-mobile-engage.com"))
        }
        val result = mockRequestModel.isMobileEngageRequest()

        result shouldBe false
    }

    @Test
    fun testIsCustomEvent_V3_true_whenItIsCustomEventV3Event() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL(EVENT_BASE))
        }
        val result = mockRequestModel.isCustomEvent()

        result shouldBe true
    }

    @Test
    fun testIsRemoteConfig_true_whenItIsRemoteConfigUrl() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL(REMOTE_CONFIG_HOST))
        }
        val result = mockRequestModel.isRemoteConfigRequest()

        result shouldBe true
    }

    @Test
    fun testIsCustomEvent_V3_false_whenItIsNotCustomEventV3Event() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL(CLIENT_BASE))
        }
        val result = mockRequestModel.isCustomEvent()

        result shouldBe false
    }

    @Test
    fun testIsRefreshContactTokenRequest_true_whenItIsRefreshContactTokenRequest() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL("$CLIENT_BASE/contact-token"))
        }
        val result = mockRequestModel.isRefreshContactTokenRequest()

        result shouldBe true
    }

    @Test
    fun testIsRefreshContactTokenRequest_false_whenItIsNotRefreshContactTokenRequest() {
        val mockRequestModel = mock(RequestModel::class.java).apply {
            whenever(url).thenReturn(URL("$CLIENT_BASE/contact"))
        }
        val result = mockRequestModel.isRefreshContactTokenRequest()

        result shouldBe false
    }

}